### ISSUE TITLE
Is it possible to check whether a specific session has been pushed into server?

### ISSUE THREAD

**Issue by charleywan2004** [bokeh/bokeh#3174]
I use the following code to authenticate a user and push and pull a session. But I found out if the "push" action takes more time, then the "pull" action just pulls a blank page, I have to refresh the browser again to get the page. So I added a wait time in between "push" and "pull". Is it possible to check whether a specific session has been pushed into server so that the "pull" action can be triggered after that? 

```
from gevent import monkey; monkey.patch_all()
from bottle import route, request, template, static_file, run, redirect
import random
import os
from subprocess import Popen
import time

@route('/login') 
def login():
    return '''
        <form action="/login" method="post">
            Username: <input name="username" type="text" />
            Password: <input name="password" type="password" />
            <input value="Login" type="submit" />
        </form>
    '''
@route('/login', method='POST')
def do_login():
    usr = request.forms.get('username')
    pwd = request.forms.get('password')

    if usr=="username" and pwd=="password":
        session_id=str(random.getrandbits(128))
        Popen('python BokehPushSession.py '+ session_id + ' ' + 'None' + ' ' + 'ws://BokehServerAddress:port/ws' )
        link = "http://BokehServerAddress:port/?bokeh-session-id="+session_id
        time.sleep(5) 
        redirect(link)
    else:
        redirect('login')

root='C:\\'
@route('/static/<filename>')
def server_static(filename):
    return static_file(filename, root=root)

run(server='gevent',host=WebServerAddress,port=port, debug=True)`

```

**Comment by havocp** [comment]
What's in BokehPushSession.py? push_session I believe does wait until the session has been pushed, the problem may be that you need to wait for your spawned BokehPushSession to finish, so `process = Popen()` then `process.wait()` or something like that (not sure of exact API). Also, in BokehPushSession be sure you `session.close()` the session from `push_session`

**Comment by charleywan2004** [comment]
In BukehPushSession.py, I have
```
session = push_session(curdoc(),sys.argv[1],None,sys.argv[2])
session.loop_until_closed()
```
I think I need to maintain infinite loop so that client can interact with server, e.g. client can adjust sliders  to update the charts in both server and client.

**Comment by havocp** [comment]
if your popen'd process is persistent I think you need some way for it to communicate with the parent process to signal that it's done with the push_session. That's more of a general question about IPC than a bokeh question. Or perhaps you could push_session in the parent process then pull_session in the spawned client process? Or find a way to avoid the client process and put all its logic in bokeh server instead?

**Comment by bryevdv** [comment]
Just to be clear I don't think this is an eventual recommended route? I think a real scenario would look like:
* running `bokeh serve my_app.py` (once) separately as part of the deploy process
* then connecting to/embedding new sessions from the flask app

or 

* creating and pushing an app to the server programmatically (once) from the flask app
* then connecting to/embedding new sessions from the flask app

Basically in general: load an app with callbacks into the bokeh server (programmatically or not), and have some way to interact with individual sessions from flask, to embed them in served templates.

Now, we may be missing some of the plumbing to make this happen quite yet. 

@havocp does this sound right? I just don't think using `popen` on a script can be our final recommended integration method.

**Comment by havocp** [comment]
No, I don't think this is how I'd do it, that's true. I was jusr assuming there's a quirky reason for the popen process in this case.

**Comment by charleywan2004** [comment]
I understand that a push session requires a session id, I have tried to pull the same session twice, it wouldn't work when I interacted with both (e.g. adjusting sliders). So I used a pair of push and pull session with the same session id. After I closed the pull session, the push session was still there. I don't know how to close it. 

Please let me know in the future how you make your recommended way happen, i.e. "
--running bokeh serve my_app.py (once) separately as part of the deploy process
--then connecting to/embedding new sessions from the flask app "

**Comment by havocp** [comment]
@charleywan2004 I don't understand enough about what you're trying to do here. Mostly I don't understand when you said "I think I need to maintain infinite loop so that client can interact with server"

Maybe this was a misunderstanding. You do not need to keep a Python client open to allow a browser client to interact with the server. I thought you meant you were interacting with the server from the Python client.

This thread may be helpful: https://groups.google.com/a/continuum.io/forum/#!topic/bokeh/xrk240aPu88

Possibly you don't need to do any push or pull session at all. If your app is in `myapp.py` then in theory you can `bokeh serve myapp.py` and then simply redirect users to `serveraddress/myapp` and they will always get a fresh session.

**Comment by bryevdv** [comment]
Right, I would imagine the most simple uses of Flask/Tornado/Django + Bokeh Server would involve one of:
* a simple redirect
* an template with `autoload_server`
This covers the case that every user gets the same exact app. 

If the app needs some kind of "customization" per user then a few ideas come to mind: 

* We could add some mechanism to pass "parameters" via the app URL. 
* Flask (or whatever) could use the python client to open  a session, manipulate it first, and then redirect/template it in a response as described above. 

The main thing we need to convey I think is that "Bokeh App is a factory for documents". The server only needs to be configured once with an app, to serve up many copies of it. 

We should definitely work out some full (simple) examples of deployments, in case there are still needed features to add.

**Comment by charleywan2004** [comment]
I tried to use "bokeh serve myApp.py" and I opened "http://localhost:5006/myApp" twice, then the second one worked (meaning I can interact with it), the first one not. I guess it does not create a copy for the first one?

**Comment by bryevdv** [comment]
@charleywan2004 Let's try to establish a baseline. I just ran `bokeh server sliders.py` on this example:

https://github.com/bokeh/bokeh/blob/master/examples/app/sliders.py

And was able to open multiple tabs and interact with all of them, using the current master as well as the latest dev build (`dev2`). Can you verify that this does or does not work for you?

**Comment by charleywan2004** [comment]
I have updated bokeh to dev2. I opened 2 "http://localhost:5006/sliders" and was only able to interact with the second one (attached is the error message).

![image](https://cloud.githubusercontent.com/assets/15861395/11381997/5af8d7f6-92cd-11e5-8837-7481499cb15c.png)

![image](https://cloud.githubusercontent.com/assets/15861395/11381952/2709f9c0-92cd-11e5-9680-5eb93eedf936.png)

**Comment by bryevdv** [comment]
That's completely bizarre. @charleywan2004 any chance you can try on other platforms?

**Comment by charleywan2004** [comment]
what other platform? I am using windows 7 professional.

I uninstalled anaconda 64 and reinstalled anaconda 32, same result.

I tried installing anaconda in another computer with windows 10, same result.

**Comment by bryevdv** [comment]
linux or osx, anything non-windows. Trying to isolate if this is windows specific. For reference, it is working for me on OSX.

**Comment by charleywan2004** [comment]
I tried Ubuntu, (installed anaconda for linux and bokeh ver 11 dev 2, the same error
![image](https://cloud.githubusercontent.com/assets/15861395/11386318/5464bf56-92ee-11e5-98b8-d31be3e68ddf.png)

**Comment by bryevdv** [comment]
I'm afraid I am at a loss, I just tried this on a Ubuntu VM and multiple sessions worked fine, including the first one, and no errors showing in console. Are you running the `bokeh` command out of the repo instead of the installed version by any chance?

**Comment by charleywan2004** [comment]
I followed the process to download anaconda from anaconda web site and updated bokeh to the latest developer version. I don't know what went wrong. Maybe it indicates some potential steps a user (like me) may miss while a developer (like you) won't.

**Comment by bryevdv** [comment]
@charleywan2004 for sure, we want to make sure that it's not easy to misuse. Can you provide the following bits of info:
* what directory you were in
* the exact command line you used
* any `PYTHONPATH`, if set
* the result of `ls bokeh*` in your relevant site-packages dir, e.g. for me:
```
$ ls ~/anaconda/lib/python3.4/site-packages/bokeh*
/Users/bryan/anaconda/lib/python3.4/site-packages/bokeh.pth
```
(yours will look different)

**Comment by mattpap** [comment]
> I tried Ubuntu, (installed anaconda for linux and bokeh ver 11 dev 2, the same error)

@bryevdv, I saw something similar on travis CI recently (just have to find the test run).

**Comment by charleywan2004** [comment]
what directory you were in
c://users/charleywan

the exact command line you used
"c://Users/charleywan/bokeh serve sliders.py"

any PYTHONPATH, if set
"C:\Users\218007366\AppData\Local\Continuum\Anaconda2;C:\Users\218007366\AppData\Local\Continuum\Anaconda2\Scripts;C:\Users\218007366\AppData\Local\Continuum\Anaconda2\Library\bin; C:\Program Files\Redis"

the result of ls bokeh* in your relevant site-packages dir, e.g. for me:
"C:\Users\charleywan\AppData\Local\Continuum\Anaconda2\Lib\site-packages\bokeh"

I updated bokeh by using "conda install -c bokeh/channel/dev bokeh"

**Comment by bryevdv** [comment]
@charleywan2004 I think we are working through this issue now in #3190, hold tight we will have a new dev build next week sometime. (ping @havocp I think this is the same issue I've been trying to run down in the PR)

**Comment by StevenCHowell** [comment]
> I have updated bokeh to dev2 and copied "sliders.py". I used "bokeh serve sliders.py" and opened 2 "http://localhost:5006/sliders" and was only able to interact with the second one (attached is the error message).

I just tried to reproduce this (Ubuntu 14.04, bokeh 0.12.4) but it worked without any issues; I could interact with the sine wave plot on multiple webpages without any errors.  Considering the amount of time that has passed, I expect this was resolved (perhaps in merged PR #3190)  and can safely be closed. 

@charleywan2004 if you are still having issues, please reopen this with additional comments.
